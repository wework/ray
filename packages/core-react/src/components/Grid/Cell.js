import React from 'react';
import PropTypes from 'prop-types';
import clsx from 'clsx';
import { settings } from '@wework/ray-core';

const breakpoints = Object.keys(settings.breakpointsInPixels);

export default function Cell({ className, push, span, ...props }) {
  const breakpointClasses = breakpoints.reduce((memo, breakpointName) => {
    const spanKey = `span-${breakpointName}`;
    if (props[spanKey]) {
      memo.push(`ray-grid__cell--span-${props[spanKey]}-${breakpointName}`);
    }

    const pushKey = `push-${breakpointName}`;
    if (props[pushKey]) {
      memo.push(`ray-grid__cell--push-${props[pushKey]}-${breakpointName}`);
    }

    return memo;
  }, []);

  const classes = clsx(
    'ray-grid__cell',
    {
      [`ray-grid__cell--span-${span}`]: span,
      [`ray-grid__cell--push-${push}`]: push
    },
    breakpointClasses,
    className
  );

  return <div className={classes} {...props} />;
}

Cell.propTypes = {
  className: PropTypes.string,
  span: PropTypes.number,
  push: PropTypes.number
};
