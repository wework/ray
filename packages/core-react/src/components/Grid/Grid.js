import React from 'react';
import PropTypes from 'prop-types';
import clsx from 'clsx';

export default function Grid({ className, ...props }) {
  const classes = clsx('ray-grid', className);

  return <div className={classes} {...props} />;
}

Grid.propTypes = {
  className: PropTypes.string
};
