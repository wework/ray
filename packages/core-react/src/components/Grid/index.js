/* eslint-disable import/prefer-default-export */
export { default as Container } from './Container';
export { default as Grid } from './Grid';
export { default as Cell } from './Cell';
