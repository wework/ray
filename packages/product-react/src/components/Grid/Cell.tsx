import clsx from 'clsx';
import React, { ReactHTML, ReactNode } from 'react';

export enum CellAlign {
  TOP = 'top',
  MIDDLE = 'middle',
  BOTTOM = 'bottom'
}

export enum CellDevice {
  DESKTOP = 'desktop',
  TABLET = 'tablet',
  PHONE = 'phone'
}

export enum CellJustify {
  TOP = 'top',
  MIDDLE = 'middle',
  BOTTOM = 'bottom'
}

export type CellProps = {
  align?: CellAlign;
  as: keyof ReactHTML;
  children?: ReactNode;
  className?: string;
  device?: CellDevice;
  full?: boolean;
  justify?: CellJustify;
  pushColumn?: 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12;
  span?: 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12;
} & typeof defaultProps;

const cellSpanClass = (span?: number, device?: CellDevice) => {
  if (span && device) {
    return 'ray-grid__cell--span-' + span + '-' + device;
  } else if (span && !device) {
    return 'ray-grid__cell--span-' + span;
  } else {
    return null;
  }
};

const cellAlignClass = (align?: CellAlign) => {
  if (align) {
    return 'ray-grid__cell--align-' + align;
  } else {
    return null;
  }
};

const cellPushClass = (pushColumn?: number, device?: CellDevice) => {
  if (pushColumn && device) {
    return 'ray-grid__cell--push-' + pushColumn + '-' + device;
  } else if (pushColumn && !device) {
    return 'ray-grid__cell--push-' + pushColumn;
  } else {
    return null;
  }
};

export const Cell = ({
  align,
  as: ComponentType,
  children,
  className,
  device,
  full,
  pushColumn,
  span,
  ...rest
}: CellProps) => (
  <ComponentType
    className={clsx(
      'ray-grid__cell',
      { 'ray-grid__cell--span-full': full },
      cellSpanClass(span, device),
      cellAlignClass(align),
      cellPushClass(pushColumn, device),
      className
    )}
    {...rest}
  >
    {children}
  </ComponentType>
);

Cell.CellAlign = CellAlign;
Cell.CellDevice = CellDevice;
Cell.CellJustify = CellJustify;

const defaultProps = {
  as: 'div'
};

Cell.defaultProps = defaultProps;

export default Cell;
