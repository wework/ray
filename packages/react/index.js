export { Container, Grid, Cell } from './src/components/Grid';
