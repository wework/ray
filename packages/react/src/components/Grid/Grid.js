import React from 'react';
import PropTypes from 'prop-types';
import clsx from 'clsx';

const Grid = ({ children, className, ...props }) => {
  const classes = clsx({
    'ray-grid': true,
    [`${className}`]: !!className
  });

  return (
    <div className={classes} {...props}>
      {children}
    </div>
  );
};

Grid.propTypes = {
  children: PropTypes.node.isRequired,
  className: PropTypes.string
};

export default Grid;
