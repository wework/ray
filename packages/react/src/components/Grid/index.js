export { default as Container } from './Container';
export { default as Grid } from './Grid';
export { default as Cell } from './Cell';
