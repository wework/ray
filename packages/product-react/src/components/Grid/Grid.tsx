import clsx from 'clsx';
import React, { ReactHTML, ReactNode } from 'react';
import Cell from './Cell';

type GridProps = {
  as?: keyof ReactHTML;
  children?: ReactNode;
  className?: string;
} & typeof defaultProps;

export const Grid = ({
  as: ComponentType,
  children,
  className,
  ...rest
}: GridProps) => {
  return (
    <ComponentType className={clsx('ray-grid', className)} {...rest}>
      {children}
    </ComponentType>
  );
};

const defaultProps = {
  as: 'div',
};

Grid.defaultProps = defaultProps;
Grid.Cell = Cell;
