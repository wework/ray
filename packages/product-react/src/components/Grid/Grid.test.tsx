import Enzyme, { configure, mount, ReactWrapper } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import * as React from 'react';
import { Grid } from './Grid';

configure({ adapter: new Adapter() });

describe('Grid', () => {
  let component: ReactWrapper<any>;

  beforeEach(() => {
    component = mount(<Grid />);
  });

  test('should match snapshot and styles', () => {
    expect(component).toMatchSnapshot();
  });

  describe('default classes', () => {
    test('should contain default class', () => {
      expect(
        component
          .find('div')
          .at(0)
          .hasClass('ray-grid'),
      ).toBe(true);
    });
    test('should contain default class & contain extra class', () => {
      component.setProps({ className: 'foo' });
      const container = component.find('div').at(0);
      expect(container.hasClass('ray-grid')).toBe(true);
      expect(container.hasClass('foo')).toBe(true);
    });
    test('should accept "as" prop as string', () => {
      component.setProps({ as: 'main' });
      expect(component.find('div').length).toBe(0);
      expect(
        component
          .find('main')
          .at(0)
          .hasClass('ray-grid'),
      ).toBe(true);
    });
  });
});
