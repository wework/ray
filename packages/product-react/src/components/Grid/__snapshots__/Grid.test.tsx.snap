// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid should match snapshot and styles 1`] = `
<Grid
  as="div"
>
  <div
    className="ray-grid"
  />
</Grid>
`;
